rust/lance-linalg/src/simd/u8.rs

```diff
@@ -0,0 +1,429 @@
+        {
+            let mut result = [0u8; 16];
+            for i in 0..16 {
+                result[i] = *ptr.add(i);
+            }
+            Self(result)
+        }
+    }
+
+    #[inline]
+    unsafe fn store(&self, ptr: *mut u8) {
+        #[cfg(target_arch = "x86_64")]
+        unsafe {
+            _mm_storeu_si128(ptr as *mut __m128i, self.0)
+        }
+        #[cfg(target_arch = "aarch64")]
+        unsafe {
+            vst1q_u8(ptr, self.0)
+        }
+        #[cfg(not(any(target_arch = "x86_64", target_arch = "aarch64")))]
+        {
+            self.store_unaligned(ptr);
+        }
+    }
+
+    #[inline]
+
+    unsafe fn store_unaligned(&self, ptr: *mut u8) {
+        #[cfg(target_arch = "x86_64")]
+        unsafe {
+            _mm_storeu_si128(ptr as *mut __m128i, self.0)
+        }
+        #[cfg(target_arch = "aarch64")]
+        unsafe {
+            vst1q_u8(ptr, self.0)
+        }
+        #[cfg(not(any(target_arch = "x86_64", target_arch = "aarch64")))]
+        {
+            for i in 0..16 {
+                *ptr.add(i) = self.0[i];
+            }
+        }
+    }
+
+    fn reduce_sum(&self) -> u8 {
+        todo!("it is not implemented yet");
+    }
+
+    #[inline]
+    fn reduce_min(&self) -> u8 {
+        #[cfg(target_arch = "x86_64")]
+        unsafe {
+            let low = _mm_and_si128(self.0, _mm_set1_epi8(0xFF_u8 as i8));
+            let high = _mm_srli_si128(self.0, 8);
+            let min_low = _mm_min_epu8(low, high);
+            let min_low = _mm_min_epu8(min_low, _mm_srli_si128(min_low, 4));
+            let min_low = _mm_min_epu8(min_low, _mm_srli_si128(min_low, 2));
+            let min_low = _mm_min_epu8(min_low, _mm_srli_si128(min_low, 1));
+            _mm_extract_epi8(min_low, 0) as u8
+        }
+        #[cfg(target_arch = "aarch64")]
+        unsafe {
+            vminvq_u8(self.0)
+        }
+        #[cfg(not(any(target_arch = "x86_64", target_arch = "aarch64")))]
+        {
+            let mut min = self.0[0];
+            for i in 1..16 {
+                min = std::cmp::min(min, self.0[i]);
+            }
+            min
+        }
+    }
+
+    #[inline]
+    fn min(&self, rhs: &Self) -> Self {
+        #[cfg(target_arch = "x86_64")]
+        unsafe {
+            Self(_mm_min_epu8(self.0, rhs.0))
+        }
+        #[cfg(target_arch = "aarch64")]
+        unsafe {
+            Self(vminq_u8(self.0, rhs.0))
+        }
+        #[cfg(not(any(target_arch = "x86_64", target_arch = "aarch64")))]
+        {
+            let mut result = [0u8; 16];
+            for i in 0..16 {
+                result[i] = std::cmp::min(self.0[i], rhs.0[i]);
+            }
+            Self(result)
+        }
+    }
+
+    fn find(&self, _val: u8) -> Option<i32> {
+        todo!()
+    }
+}
+
+impl Shuffle for u8x16 {
+    fn shuffle(&self, indices: u8x16) -> Self {
+        #[cfg(target_arch = "x86_64")]
+        unsafe {
+            Self(_mm_shuffle_epi8(self.0, indices.0))
```

Comment:
Would it be faster if we can use https://doc.rust-lang.org/beta/core/arch/x86_64/fn._mm256_shuffle_epi8.html (u8x32)

Reply:
Yeah I believe so,
Chose u8x16 because it fit in arm register size

Reply:
we can implement `u8x32` as 2 of 128bit register on arm? Just in general this can speed up x86 old cpu a lot, similar to https://github.com/lancedb/lance/blob/main/rust/lance-linalg/src/simd/f32.rs#L462

Reply:
doable, let's do this next PR? that would need to change the computation logic as well, because there are only 16 centroids in distance_table for each sub vector.